### ISSUE TITLE
Altering a colum to an enum with useNative: true fails, tries to create enum twice

### ISSUE THREAD

**Issue by mogzol** [knex/knex#3657]
# Environment

Knex version: 0.20.9
Database + version: Postgres 12
OS: Linux (Ubuntu)

# Bug

If you attempt to alter a column to change it's type to an enum, and you use the `useNative: true` enum option, knex will try to create the enum twice, which causes a "type already exists" error.

Here is an example migration which demonstrates the problem:

```typescript
export async function up(knex: Knex): Promise<any> {
  return knex.schema
    .debug(true)
    .createTable("orders", (table) => {
      table.text("drink");
    })
    .alterTable("orders", (table) => {
      // This one works
      table.enum("size", ["small", "medium", "large"], {
        enumName: "drink_sizes",
        useNative: true,
      });

      // This one tries to create the enum twice
      table
        .enum("drink", ["coffee", "tea", "water"], {
          enumName: "drink_types",
          useNative: true,
        })
        .alter();
    });
}
```

If you look at the debug output from that migration, you will see that the first enum in the alterTable part generates the sql fine, but the second one (using `.alter()`) will create two identical `create type` sql statements, which causes Postgres to fail with 'type "drink_types" already exists'

EDIT: 

tl;dr Altering table to enum seems to try to create column type twice:

https://runkit.com/embed/xemoadwjsojw

**Comment by elhigu** [comment]
`.alter()` basically doesn't know any information, what was the original type of the column that you are changing to be `enum` so it doesn't have any information about for example foreign keys / enum types etc. that should be removed.

Even `existingType: true` probably doesn't work in this case, because it was added for slightly different scenario https://github.com/knex/knex/pull/2719

I suppose this could be implemented with another same kind of attribute `deleteOldEnumType : true` which explicitly tells knex that this enum was actually previously declared and the old typ should be removed before creating a new enum type for that enumName.

**Comment by elhigu** [comment]
I changed this from bug -> feature request, since knex actually work in this case exactly as expected.

**Comment by mogzol** [comment]
But in this scenario I wouldn't expect alter to delete the existing type, so that's fine, I just don't see why it is trying to create the type twice? This essentially makes it impossible to use `.enum()` with `.alter()` if you're using native enums.

Is there any scenario where `.alter()` **should** create the type? It seems like the `.enum()` part is responsible for that, so why does alter also try to do it?

**Comment by elhigu** [comment]
> But in this scenario I wouldn't expect alter to delete the existing type, so that's fine, I just don't see why it is trying to create the type twice? 

So... you are not changing the list of valid enum values and those have been `["coffee", "tea", "water"]` all along? 

It is trying to recreate enum type, because knex does not know that enum type has already been created earlier. In that case for you `existingType: true` would work, since it tells to knex that enum type (with name `drink_types `) was already previously declared.

For example if you are altering `string` type of column to be native enum then you should not add  `existingType: true` to make sure that knex actually creates the type before using it.

> This essentially makes it impossible to use .enum() with .alter() if you're using native enums.

No. You can still alter default value of enum and to convert for example string column to native enum. But it essentially makes impossible to update enum type to have more valid enum values.

Also if you convert column to be something else afterwards, you might need to cleanup enum type separately with raw sql.

**Comment by mogzol** [comment]
>No. You can still alter default value of enum and to convert for example string column to native enum. But it essentially makes impossible to update enum type to have more valid enum values.

This is what I am trying to do. In the example I posted I initially create `drink` as a `text` column, and then try to alter it to an `enum`. This causes the error described.

The issue seems to be that `ColumnCompiler.getColumnType()` (which add the `create type` query) gets called twice for `.enum().alter()`.

First it is called while building alterColumns here: https://github.com/knex/knex/blob/92d394432cef729d1b434a898e27163677c72b07/lib/schema/tablecompiler.js#L79

And then it is called again during the `alterColumns` call a few lines later, which in turn calls `addColumns` in the Postgres tablecompiler, which gets the type again: https://github.com/knex/knex/blob/92d394432cef729d1b434a898e27163677c72b07/lib/dialects/postgres/schema/tablecompiler.js#L59

**Comment by mogzol** [comment]
So maybe a simple flag inside the `ColumnCompiler` instance to keep track of whether the `create type` query has already been added would resolve this? Then only the first call to `getColumnType()` would add the query?

**Comment by elhigu** [comment]
True... there seems to be bug in that:

https://runkit.com/embed/xemoadwjsojw

It really tries to create column type twice 👍

**Comment by elhigu** [comment]
> So maybe a simple flag inside the ColumnCompiler instance to keep track of whether the create type query has already been added would resolve this? Then only the first call to getColumnType() would add the query?

It sounds like a bad idea that `getColumnType()` creates that query as a side effect. Without looking the code to me it would make more sense to have createColumnType() and getColumnType() as a separate methods.

**Comment by mogzol** [comment]
Ah didn't see that comment. Well then I can update that MR once I have some free time later, unless someone else wants to do it

**Comment by kibertoad** [comment]
Released in 0.95.0
